dbms/src/Storages/StorageFile.cpp

```diff
@@ -39,6 +47,47 @@ namespace ErrorCodes
     extern const int EMPTY_LIST_OF_COLUMNS_PASSED;
 }
 
+namespace
+{
+
+Strings LSWithRegexpMatching(const String & path_for_ls, const String & for_match)
+{
+    Strings result;
+    size_t first_glob = for_match.find_first_of("*?{");
+
+    size_t end_of_path_without_globs = for_match.substr(0, first_glob).rfind('/');
+    String path_with_globs = for_match.substr(end_of_path_without_globs);   /// begin with '/'
+    String path_without_globs = path_for_ls + for_match.substr(1, end_of_path_without_globs); /// ends with '/'
```

Comment:
Maybe something like `suffix_with_globs` and `prefix_without_globs`, to make it clear that these are two parts of the same path, and not some unrelated paths.